[ShellScript] scope colon as a path separator inside quoted strings

keith-hall opened the pull request:
This makes it much easier to visualize e.g. setting the `PATH` env var in scripts because the `:` can be highlighted as a separate color to the rest of the string.

jfcherng commented:
```sh
export SOMETHING=/etc/test:/var/test
#                         ^ punctuation.separator.sequence
```
should pass

deathaxe commented:
We've used `punctuation.separator.path` for path segment separaturs (slashes), only so far - not as separator for a list of paths. Hence I'd also vote for `punctuation.separator.sequence`.

As colons might appear everywhere in arbritary strings with any kind of meaning, I am however not sure whether we should make any assumption about the meaning of certain tokens/chars in a string, beyound escapes and placeholders.

keith-hall commented:
I believe I have addressed the suggestions.

> As colons might appear everywhere in arbritary strings with any kind of meaning, I am however not sure whether we should make any assumption about the meaning of certain tokens/chars in a string, beyound escapes and placeholders.

I generally agree, but in this case I believe the usefulness of scoping common patterns outweighs the "pure syntax" perspective. But if we vote to not have this, I guess I will need to maintain a local override or extended syntax which I was hoping to avoid as I thought this could benefit everyone...

deathaxe commented on ShellScript/Bash.sublime-syntax:
```diff
@@ -1991,6 +1995,10 @@ contexts:
       scope: constant.character.escape.shell
 
+  string-commons:
+    - match: ':(?!$|\s|//)'
+      scope: punctuation.separator.sequence.shell
```
Having a `string-commons` next to `string-prototype` feels a bit duplicate.

I'd therefore suggest to 

1. either rename it to something which expresses the meaning of the contained pattern more precisly 
2. or to move this pattern to `string-prototype`. It was added as empty map as there was no content yet. It needs to be extended via `- meta_prepend` by extending syntaxes anyway.

2nd solution would also add highlighting of path separators in ansi-c strings but exclude it from heredocs. Not sure about ansi-c strings, but I'd probably would not highlight colons in heredocs as those may contain any kind of larger text blocks with higher risk of false positives, woudln't they?

keith-hall replied:
agreed with all your points, so I went for the 1st solution of renaming the new context :)

FichteFoll commented on ShellScript/Bash.sublime-syntax:
```diff
@@ -1991,6 +1995,10 @@ contexts:
       scope: constant.character.escape.shell
 
+  string-path-separators:
+    - match: ':(?!$|\s|//)'
```
```suggestion
    - match: ':(?=\$|/)'
```
Perhaps this would be a decent approximation for a list of allowed characters after a colon when combining multiple paths together? It should be very rare to combine two relative paths.

deathaxe replied:
Theoretical should also handle `Class::method()`. There may be various possible false positives anyway.

michaelblyons replied:
Found another accidental match in `'([^.[:space:]]+)'`

keith-hall replied:
applied FichteFoll's suggestion, and added syntax tests to prove those examples are not scoped/treated as path separators

deathaxe replied:
May something like `/path/:~/dir` be something to considder? If yes, ... `:(?=\$|~?/(?!/))`

FichteFoll replied:
And the current (or a relative) directory `/path:.:../sibling`.

Can't think of anything else right now.

keith-hall replied:
done

FichteFoll replied:
Not sure why you require `.` and `~` to be followed by a slash, which is not a technical requirement, but it should be unlikely enough to not be relevant in practice, at least for `~`. `.` I could see but is probably also equivalent to an empty segment (which I remember seeing sometimes but probably way back on Windows).

Current changes are fine imo, but feel free to revise.

deathaxe replied:
You mean something like ...

```
- match: ':(?=\$|~|\.\.?|/(?!/))'
```

I haden't `bla:~:bla` in mind when writing my proposal.

keith-hall replied:
The current regex suits my personal use case, and I'm worried that making it more lenient could cause more false positives. My thinking would be to merge as is, and tweak it based on feedback later. But if we want, we can change it now also, I don't mind, just need someone to make the decision :)

deathaxe replied:
I made it for you ;-) as somehow sharing concerns about false positives.

deathaxe commented on ShellScript/Bash.sublime-syntax:
```diff
@@ -1332,7 +1332,7 @@ contexts:
     # characters ‘\’, ‘$’, and ‘`’.
     - include: string-escapes
     - include: string-interpolations
-    - include: string-commons
```
Do we really want to highlight it in heredocs?

If yes, should probably also include the other kinds below.

keith-hall replied:
probably we don't, I have now removed it from here :)
